calico_cni/calico_cni.py

```diff
@@ -245,16 +246,17 @@ def delete(self):
 
         _log.info("Finished removing container: %s", self.container_id)
 
-    def _assign_ip(self):
+    def _assign_ip(self, env):
         """Assigns and returns an IPv4 address using the IPAM plugin
         specified in the network config file.
 
         :return: IPAddress - The assigned IP address.
         """
+        assert env[CNI_COMMAND_ENV] == CNI_CMD_ADD
         # Call the IPAM plugin.  Returns the plugin returncode,
         # as well as the CNI result from stdout.
         _log.info("Assigning IP address")
-        rc, result = self._call_ipam_plugin()
+        rc, result = self._call_ipam_plugin(env)
 
         if rc:
             # The IPAM plugin failed to assign an IP address. At this point in 
```

Comment:
I think you need to move this `if rc:` block down after the `json.loads(result)` section, and do the error handling there.  We should be using the return code to determine if an error has occured - something like this.

```
if rc:
  code = self.ipam_response.get("code", ERROR_CODE_GENERIC)
  msg = self.ipam_response.get("msg", "Unknown IPAM error")
  details = self.ipam_response.get("details")
  self._print_error_response(code, msg, details)
  sys.exit(int(code))
```